src/NuGet.Clients/NuGet.SolutionRestoreManager/SolutionRestoreWorker.cs

```diff
@@ -208,9 +229,15 @@ private void SolutionEvents_AfterClosing()
             // on-board request onto pending restore operation
             _pendingRequests.Value.TryAdd(request);
 
+            // Await completion of the requested restore operation.
+            // The caller will be unblocked immediately upon
+            // cancellation request via provided token.
+            // Method returns false in case of cancellation is requested.
+            var cancellationTcs = new TaskCompletionSource<bool>();
+            using (token.Register(() => cancellationTcs.SetResult(false)))
             using (_joinableCollection.Join())
```

Comment:
Rather than awaiting raw `Task` inside a Join block, I once again recommend making `pendingRestore` a `JoinableTask` and then you can simply await it. This change may not be something you want for RTW unless you're confident in it, but I recommend you do it for 15.1 when you have a bit of time to stabilize. The risk mitigation wouldn't normally be required but if you're relying on throwing arbitrary things into your `JoinableTaskCollection` that you may depend on (as a way to avoid following the threading rules too closely) then making this pattern enhancement would show up as deadlocks till you tighten up your compliance.